core/client/fs/src/main/java/alluxio/client/metrics/ClientMasterSync.java

```diff
@@ -33,41 +38,75 @@
+  private final AlluxioConfiguration mConf;
 
   /**
    * Constructs a new {@link ClientMasterSync}.
    *
    * @param masterClient the master client
-   * @param appId the app ID for this client
+   * @param conf Alluxio configuration
    */
-  public ClientMasterSync(MetricsMasterClient masterClient, String appId) {
+  public ClientMasterSync(MetricsMasterClient masterClient, AlluxioConfiguration conf) {
     mMasterClient = Preconditions.checkNotNull(masterClient, "masterClient");
-    mAppId = Preconditions.checkNotNull(appId, "appId");
+    mApplicationId = new ConcurrentHashSet<>();
+    mConf = conf;
   }
 
-  @Override
-  public synchronized void heartbeat() throws InterruptedException {
-    List<alluxio.grpc.Metric> metrics = new ArrayList<>();
-    for (Metric metric : MetricsSystem.allClientMetrics()) {
-      metric.setInstanceId(mAppId);
-      metrics.add(metric.toProto());
-    }
-    try {
-      mMasterClient.heartbeat(metrics);
-    } catch (IOException e) {
-      // An error occurred, log and ignore it or error if heartbeat timeout is reached.
-      LOG.error("Failed to heartbeat to the metrics master:", e);
-      mMasterClient.disconnect();
-    }
+  /**
+   * Adds a new application Id to be included in the metrics heartbeat.
+   *
+   * @param appId the application identifier to be added
+   */
+  public void addAppId(String appId) {
+    Preconditions.checkNotNull(appId);
+    Preconditions.checkArgument(!appId.equals(""));
+    mApplicationId.add(appId);
   }
 
-  @Override
-  public void close() {
+  /**
+   * Removes and application Id from being included in the metrics heartbeat.
+   *
+   * @param appId the application identifier to be removed
+   */
+  public void removeAppId(String appId) {
+    mApplicationId.remove(appId);
+  }
+
+  /**
+   * Sends metrics to the master for each application Id that has been added.
+   */
+  public synchronized void heartbeat() {
+    // Only perform the heartbeat if there are applications to track
+    if (mApplicationId.size() > 0) {
+      List<alluxio.grpc.ClientMetrics> clientMetrics = new ArrayList<>();
+      List<Metric> allClientMetrics = MetricsSystem.allClientMetrics();
+      String hostname = NetworkAddressUtils.getClientHostName(mConf);
+      for (String appId : mApplicationId) {
+        List<alluxio.grpc.Metric> metrics = new ArrayList<>();
+        for (Metric metric : allClientMetrics) {
+          metric.setInstanceId(appId);
+          metrics.add(metric.toProto());
+        }
+        clientMetrics.add(ClientMetrics.newBuilder()
+            .setHostname(hostname)
+            .setClientId(appId)
+            .addAllMetrics(metrics)
+            .build());
+      }
+      try {
+        mMasterClient.heartbeat(clientMetrics);
+      } catch (IOException e) {
+        // WARN instead of ERROR as metrics are not critical to the application function
+        LOG.warn("Failed to send metrics to master: ", e);
+        mMasterClient.disconnect();
```

Comment:
Move the disconnect logic to the call to `heartbeat`?

Reply:
Sorry I don't quite understand what you mean

Reply:
The `MetricsMasterClient` is responsible for connecting on each `heartbeat` call, but it does not clean up (disconnect) on any network errors and instead relies on the caller to explicitly disconnect (like what you are doing here). The `heartbeat` call should make the decision on when to disconnect before throwing an exception (since some exceptions could be unrelated to the connection).
I looked at some of the other clients and we don't have a strict convention (most others use `retryRPC` so this problem is avoided).